connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1020,7 +1107,9 @@ public HerderRequest(long at, Callable<Void> action, Callback<Void> callback) {
 
         @Override
         public int compareTo(HerderRequest o) {
-            return Long.compare(at, o.at);
+            final int soonest = Long.compare(at, o.at);
+            // If tied, returning a positive value should respect insertion order.
+            return soonest != 0 ? soonest : 1;
```

Comment:
Strictly speaking I agree about how someone needs to implement `compareTo`. But for this specific case we don't have to. 

I should have described the above as "Best effort insertion order" :)

This is a private inner class used only by the herder. It's not meant to be used with other collections outside this class. Thus, we can make an implementation specific optimization and resolve the tie this way, without using an atomic integer as a counter. 

Anyone who messes with this class needs anyway to observe that `compareTo` depends only on the comparison of timestamps at a millisecond resolution (I know that the actual granularity  depends on the OS and can be even greater). 

That means that completely different requests can be compared as equal anyways. But since the usage of this class is very restricted we can optimize as far as it makes sense. So observing that returning a positive integer when timestamps are equal actually preserves insertion order is not bad here, IMHO.